File: objectiveai-py/objectiveai/functions/profiles/computations/http.py
"""HTTP functions for profile computation endpoints."""

from __future__ import annotations

from typing import TYPE_CHECKING, Union

if TYPE_CHECKING:
    from objectiveai.client import ObjectiveAI
    from objectiveai.functions.profiles.computations.request import (
        FunctionProfileComputationCreateParams,
    )
    from objectiveai.functions.profiles.computations.response.streaming import (
        FunctionProfileComputationChunk,
    )
    from objectiveai.functions.profiles.computations.response.unary import (
        FunctionProfileComputation,
    )
    from objectiveai.stream import Stream


async def compute_profile(
    client: ObjectiveAI,
    params: FunctionProfileComputationCreateParams,
    *,
    fremote: str | None = None,
    fowner: str | None = None,
    frepository: str | None = None,
    fcommit: str | None = None,
) -> Union[FunctionProfileComputation, Stream[FunctionProfileComputationChunk]]:
    """Compute a profile for a function.

    Path is built from the remote source arguments. If none are provided,
    uses the inline computation path. If ``params.stream`` is true, returns
    a streaming response.
    """
    if fremote and fowner and frepository:
        parts = [f"functions/{fremote}/{fowner}/{frepository}"]
        if fcommit:
            parts.append(fcommit)
        parts.append("profiles/compute")
        path = "/".join(parts)
    else:
        path = "functions/profiles/compute"

    if getattr(params, "stream", None):
        return await client.post_streaming(path, params)
    return await client.post_unary(path, params)

File: objectiveai-js/src/functions/profiles/computations/http.ts
import { ObjectiveAI, type RequestOptions } from "../../../client";
import { Stream } from "../../../stream";
import type { FunctionsProfilesComputationsRequestRequest } from "./request/request";
import type { FunctionsProfilesComputationsResponseUnaryFunctionProfileComputation } from "./response/unary/functionProfileComputation";
import type { FunctionsProfilesComputationsResponseStreamingFunctionProfileComputationChunk } from "./response/streaming/functionProfileComputationChunk";

function buildComputationPath(request: FunctionsProfilesComputationsRequestRequest): string {
  if (!("path" in request)) {
    return "/functions/profiles/compute";
  }
  const { path } = request;
  let url = `/functions/${path.fremote}/${path.fowner}/${path.frepository}`;
  if (path.fcommit != null) url += `/${path.fcommit}`;
  url += "/profiles/compute";
  return url;
}

export function functionsProfilesComputationsComputeProfile(
  client: ObjectiveAI,
  request: FunctionsProfilesComputationsRequestRequest & { body: { stream: true } },
  options?: RequestOptions,
): Promise<Stream<FunctionsProfilesComputationsResponseStreamingFunctionProfileComputationChunk>>;
export function functionsProfilesComputationsComputeProfile(
  client: ObjectiveAI,
  request: FunctionsProfilesComputationsRequestRequest & { body: { stream?: false | null } },
  options?: RequestOptions,
): Promise<FunctionsProfilesComputationsResponseUnaryFunctionProfileComputation>;
export function functionsProfilesComputationsComputeProfile(
  client: ObjectiveAI,
  request: FunctionsProfilesComputationsRequestRequest,
  options?: RequestOptions,
): Promise<
  | Stream<FunctionsProfilesComputationsResponseStreamingFunctionProfileComputationChunk>
  | FunctionsProfilesComputationsResponseUnaryFunctionProfileComputation
> {
  const path = buildComputationPath(request);
  if (request.body.stream) {
    return client.post_streaming<FunctionsProfilesComputationsResponseStreamingFunctionProfileComputationChunk>(
      path,
      request.body,
      options,
    );
  }
  return client.post_unary<FunctionsProfilesComputationsResponseUnaryFunctionProfileComputation>(
    path,
    request.body,
    options,
  );
}
